profiler/profile.go

```diff
@@ -216,6 +217,22 @@ func (b *batch) addProfile(p *profile) {
 }
 
 func (p *profiler) runProfile(pt ProfileType) ([]*profile, error) {
+	var extra []*pprofile.Profile
+	if pt == HeapProfile {
+		// For the heap profile, we'd also like to include C allocations
+		// if that extension is enabled and have the allocations show up
+		// in the same profile
+		//
+		// TODO: Support non-delta profiles for C allocations?
+		if cAlloc, ok := extensions.GetCAllocationProfiler(); ok && p.cfg.deltaProfiles {
+			cAlloc.Start(2 * 1024 * 1024)
+			p.interruptibleSleep(p.cfg.period)
```

Comment:
I think runProfile() is done sequentially, so if we do an `interruptibleSleep` here, it means we won't be able to run the c allocation profiles concurrently with the CPU profiler? If yes, that's probably an issue we need to figure out.

Reply:
Agreed, I have another PR in the works to run the profiles concurrently. I found some race detector warnings when implementing it that are addressed by #1259, and I want to have those in place before adding concurrency. For now there's nothing actually _implementing_ the C allocation profiler in this repo so this change won't actually block anything yet. If you'd prefer, though, we can add concurrency first and then merge this second.

Reply:
I'm okay with the concurrency issue being addressed in a follow-up patch.

Another issue I just realized is that this kind of breaks the abstraction of the different profile types.

Would it be possible to integrate this code into the definition of the heap profile type? I realize this might need changes to that interface, but it seems like it might be worth it?

https://github.com/DataDog/dd-trace-go/blob/c7e6cbedc3880706b2ed77be5feb0f67b0b9706f/profiler/profile.go#L104-L111

Reply:
Yeah, I think it might be worth it. I kind of took the path of least resistance here: The current abstraction has `profileType.Collect` returning a serialized profile. But if I want to merge C allocations into the heap profile, I'd want the C allocation profile to be in a format that's easy to merge. So that means a `*profile.Profile` as opposed to encoded in pprof format(*). But the heap profile is only a `*profile.Profile` while the delta profile is computed, so the easiest thing to do was to add a pathway from collecting the profiles to computing the delta where I could pass the profile to merge.

But despite this being the smallest change I could think of to accomplish my goal, I agree that it sort of starts to break the abstraction. I can take a stab at doing a refactor to make this change nicer.

_(*) Side note: pprofs are protocol buffers, and protocol buffers are just concatenated sequences of tag-value records. It would be cool to just have the C allocation samples already encoded and just append them to the end of an existing pprof. Likewise, it would be cool to do delta profiles by modifying a pprof in-place, or at least without having to fully decode it into a `*profile.Profile`. But that would be significantly more complicated. Maybe a good research week project?_

Reply:
Okay, I pushed another commit that pulls the heap profile special case closer to where the profile is collected. The code could probably still use a bigger refactor but hopefully this is at least good enough for this PR.